From 7ccdb828aeeab8ed6b24e37d22efd40f1a7c96d7 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Mon, 3 Aug 2026 13:35:08 +0100
Subject: [PATCH] mlcustomize: Add new --selinux-relabel-excludes flag

https://github.com/libguestfs/libguestfs/commit/f5ad08415182cd7ecac376aad19ffaa81bec7503

Update m4/guestfs-libraries.m4 to make libguestfs 1.61.1 required.
This is for the new guestfs_setfiles 'excludes' parameter, added in:

https://github.com/libguestfs/libguestfs/commit/5364e6abbcd1ee42f23b26306225269329e7c663

Update the common submodule, bringing in:

  Richard W.M. Jones (3):
        mlcustomize/customize_cmdline.mli: Fix differences in generated code
        mlcustomize: Update generated files
        mlcustomize: Pass --selinux-relabel-excludes through to setfiles

(cherry picked from commit 4dcf8ea13ca2516005cdb64edbbb2534fe612d49)
---
 common                  | 2 +-
 m4/guestfs-libraries.m4 | 5 +++--
 2 files changed, 4 insertions(+), 3 deletions(-)

Submodule common 9584c82f..9c02b103:
diff --git a/common/mlcustomize/SELinux_relabel.ml b/common/mlcustomize/SELinux_relabel.ml
index 62450e41..3fd97221 100644
--- a/common/mlcustomize/SELinux_relabel.ml
+++ b/common/mlcustomize/SELinux_relabel.ml
@@ -28,11 +28,11 @@ module G = Guestfs
  * [libguestfs.git/daemon/selinux.ml].
  *)
 
-let rec relabel (g : G.guestfs) =
+let rec relabel (g : G.guestfs) excludes =
   (* Is the guest using SELinux?  (Otherwise this is a no-op). *)
   if is_selinux_guest g then (
     try
-      use_setfiles g;
+      use_setfiles g excludes;
       (* That worked, so we don't need to autorelabel. *)
       g#rm_f "/.autorelabel"
     with Failure _ ->
@@ -47,7 +47,7 @@ and is_selinux_guest g =
   g#is_file ~followsymlinks:true "/usr/sbin/load_policy" &&
   g#is_file ~followsymlinks:true "/etc/selinux/config"
 
-and use_setfiles g =
+and use_setfiles g excludes =
   (* Is setfiles / SELinux relabelling functionality available? *)
   if not (g#feature_available [| "selinuxrelabel" |]) then
     failwith "no selinux relabel feature";
@@ -121,5 +121,8 @@ and use_setfiles g =
             List.sort compare |>  (* sort them for consistency *)
             Array.of_list in
 
+  (* Excludes is actually a string array, not string list. *)
+  let excludes = Array.of_list excludes in
+
   (* Relabel everything. *)
-  g#setfiles specfile mps
+  g#setfiles ~excludes specfile mps
diff --git a/common/mlcustomize/SELinux_relabel.mli b/common/mlcustomize/SELinux_relabel.mli
index 7b4f7ff2..5f1ae7ff 100644
--- a/common/mlcustomize/SELinux_relabel.mli
+++ b/common/mlcustomize/SELinux_relabel.mli
@@ -18,12 +18,19 @@
 
 (** SELinux-relabel the filesystem. *)
 
-val relabel : Guestfs.guestfs -> unit
+val relabel : Guestfs.guestfs -> string list -> unit
 (** Relabel the mounted guestfs filesystem using the current SELinux
     policy that applies to the guest.
 
     If the guest does not look like it uses SELinux, this does nothing.
 
+    The parameters are the guestfs handle and a list of directories
+    that may be excluded (by passing I<--selinux-relabel-exclude>
+    one or more times).
+
     In case relabelling is not possible (since it is an optional
     feature which requires the setfiles(8) program), instead we
-    fall back to touching [/.autorelabel]. *)
+    fall back to touching [/.autorelabel].
+
+    The excludes list only affects setfiles relabelling,
+    not [/.autorelabel] since we do not control how that works. *)
diff --git a/common/mlcustomize/customize-options.pod b/common/mlcustomize/customize-options.pod
index 95767cc7..f6e68f1e 100644
--- a/common/mlcustomize/customize-options.pod
+++ b/common/mlcustomize/customize-options.pod
@@ -386,6 +386,22 @@ It cannot delete directories, only regular files.
 
 This is a compatibility option that does nothing.
 
+=item B<--selinux-relabel-exclude> DIR
+
+Exclude directories from being relabelled.
+
+This advanced option lets you list directories in the guest which
+should not be relabelled, even when SELinux relabelling is
+enabled.  Use this carefully, as any changes that are made
+inside these directories during customization will have incorrect
+SELinux labels, leading to potential failures later, so you must
+be sure that the directories do not need relabelling.
+
+If in doubt, do not use this option.
+
+You can pass the option multiple times, eg.
+I<--selinux-relabel-exclude=/foo> I<--selinux-relabel-exclude=/bar>
+
 =item B<--ssh-inject> USER[:SELECTOR]
 
 Inject an ssh key so the given C<USER> will be able to log in over
diff --git a/common/mlcustomize/customize-synopsis.pod b/common/mlcustomize/customize-synopsis.pod
index f5eeb69d..d6c4a280 100644
--- a/common/mlcustomize/customize-synopsis.pod
+++ b/common/mlcustomize/customize-synopsis.pod
@@ -14,4 +14,4 @@
     [--truncate-recursive PATH] [--uninstall PKG,PKG..] [--update]
     [--upload FILE:DEST] [--write FILE:CONTENT] [--no-logfile]
     [--password-crypto md5|sha256|sha512] [--no-selinux-relabel]
-    [--selinux-relabel]
+    [--selinux-relabel] [--selinux-relabel-exclude DIR]
diff --git a/common/mlcustomize/customize_cmdline.ml b/common/mlcustomize/customize_cmdline.ml
index b8eda519..24e05dc9 100644
--- a/common/mlcustomize/customize_cmdline.ml
+++ b/common/mlcustomize/customize_cmdline.ml
@@ -115,6 +115,8 @@ and flags = {
       (* --no-selinux-relabel *)
   selinux_relabel_ignored : bool;
       (* --selinux-relabel *)
+  selinux_relabel_excludes : string list;
+      (* --selinux-relabel-exclude *)
 }
 
 type argspec = Getopt.keys * Getopt.spec * Getopt.doc
@@ -125,6 +127,7 @@ let rec argspec ?(v2v = false) () =
   let password_crypto = ref None in
   let no_selinux_relabel = ref false in
   let selinux_relabel_ignored = ref false in
+  let selinux_relabel_excludes = ref [] in
 
   let rec get_ops () = {
     ops = List.rev !ops;
@@ -135,6 +138,7 @@ let rec argspec ?(v2v = false) () =
     password_crypto = !password_crypto;
     no_selinux_relabel = !no_selinux_relabel;
     selinux_relabel_ignored = !selinux_relabel_ignored;
+    selinux_relabel_excludes = !selinux_relabel_excludes;
   }
   in
 
@@ -488,6 +492,15 @@ let rec argspec ?(v2v = false) () =
       s_"Compatibility option doing nothing"
     ),
     None, "This is a compatibility option that does nothing.", false;
+    (
+      [ L"selinux-relabel-exclude" ],
+      Getopt.String (
+        s_"DIR",
+        List.push_back selinux_relabel_excludes
+      ),
+      s_"Exclude directories from SELinux relabelling"
+    ),
+    Some "DIR", "Exclude directories from being relabelled.\n\nThis advanced option lets you list directories in the guest which\nshould not be relabelled, even when SELinux relabelling is\nenabled.  Use this carefully, as any changes that are made\ninside these directories during customization will have incorrect\nSELinux labels, leading to potential failures later, so you must\nbe sure that the directories do not need relabelling.\n\nIf in doubt, do not use this option.\n\nYou can pass the option multiple times, eg.\nI<--selinux-relabel-exclude=/foo> I<--selinux-relabel-exclude=/bar>", false;
   ]
   and customize_read_from_file filename =
     let forbidden_commands = [
diff --git a/common/mlcustomize/customize_cmdline.mli b/common/mlcustomize/customize_cmdline.mli
index baacda9d..abe10953 100644
--- a/common/mlcustomize/customize_cmdline.mli
+++ b/common/mlcustomize/customize_cmdline.mli
@@ -107,17 +107,24 @@ and flags = {
       (* --no-selinux-relabel *)
   selinux_relabel_ignored : bool;
       (* --selinux-relabel *)
+  selinux_relabel_excludes : string list;
+      (* --selinux-relabel-exclude *)
 }
 
 type argspec = Getopt.keys * Getopt.spec * Getopt.doc
 val argspec : ?v2v:bool -> unit -> (argspec * string option * string) list * (unit -> ops)
-(** This returns a pair [(list, get_ops)].
+(** Return the argspec (a list of arguments, see {!Getopt} module)
+    with the full list of virt-customize options.
+
+    If the parameter [~v2v] is true then this excludes parameters
+    that should be excluded from virt-v2v.  For example
+    [--inject-qemu-ga] is excluded from the list because virt-v2v
+    decides itself whether and how to inject the QEMU guest agent.
+
+    This returns a pair [(list, get_ops)].
 
     [list] is a list of the command line arguments, plus some extra data.
 
     [get_ops] is a function you can call {i after} command line parsing
     which will return the actual operations specified by the user on the
-    command line.
-
-    If the parameter [~v2v] is true then this excludes parameters
-    that should be excluded from virt-v2v. *)
\ No newline at end of file
+    command line.  This can be passed to {!Customize_run.run} *)
\ No newline at end of file
diff --git a/common/mlcustomize/customize_run.ml b/common/mlcustomize/customize_run.ml
index 1baf0f3c..4878c00f 100644
--- a/common/mlcustomize/customize_run.ml
+++ b/common/mlcustomize/customize_run.ml
@@ -361,7 +361,7 @@ let run (g : G.guestfs) root (ops : ops) =
 
   if not ops.flags.no_selinux_relabel then (
     message (f_"SELinux relabelling");
-    SELinux_relabel.relabel g
+    SELinux_relabel.relabel g ops.flags.selinux_relabel_excludes
   );
 
   (* Clean up the log file:
diff --git a/common/mlcustomize/v2v-customize-options.pod b/common/mlcustomize/v2v-customize-options.pod
index 6252ffe9..fee5f984 100644
--- a/common/mlcustomize/v2v-customize-options.pod
+++ b/common/mlcustomize/v2v-customize-options.pod
@@ -332,6 +332,22 @@ It cannot delete directories, only regular files.
 
 This is a compatibility option that does nothing.
 
+=item B<--selinux-relabel-exclude> DIR
+
+Exclude directories from being relabelled.
+
+This advanced option lets you list directories in the guest which
+should not be relabelled, even when SELinux relabelling is
+enabled.  Use this carefully, as any changes that are made
+inside these directories during customization will have incorrect
+SELinux labels, leading to potential failures later, so you must
+be sure that the directories do not need relabelling.
+
+If in doubt, do not use this option.
+
+You can pass the option multiple times, eg.
+I<--selinux-relabel-exclude=/foo> I<--selinux-relabel-exclude=/bar>
+
 =item B<--ssh-inject> USER[:SELECTOR]
 
 Inject an ssh key so the given C<USER> will be able to log in over
diff --git a/common/mlcustomize/v2v-customize-synopsis.pod b/common/mlcustomize/v2v-customize-synopsis.pod
index 4c1a1547..5a1bc009 100644
--- a/common/mlcustomize/v2v-customize-synopsis.pod
+++ b/common/mlcustomize/v2v-customize-synopsis.pod
@@ -12,4 +12,4 @@
     [--truncate-recursive PATH] [--uninstall PKG,PKG..] [--update]
     [--upload FILE:DEST] [--write FILE:CONTENT] [--no-logfile]
     [--password-crypto md5|sha256|sha512] [--no-selinux-relabel]
-    [--selinux-relabel]
+    [--selinux-relabel] [--selinux-relabel-exclude DIR]
diff --git a/m4/guestfs-libraries.m4 b/m4/guestfs-libraries.m4
index 22af5db7..6334647f 100644
--- a/m4/guestfs-libraries.m4
+++ b/m4/guestfs-libraries.m4
@@ -19,8 +19,9 @@ dnl Any C libraries required by virt-v2v.
 
 dnl Of course we need libguestfs.
 dnl
-dnl We need libguestfs >= 1.58.1-6.el10 for guestfs_xfs_info2
-dnl and for text: and base64: prefix in LUKS funcs.
+dnl We need libguestfs >= 1.58.1-9.el10 for guestfs_xfs_info2
+dnl and for text: and base64: prefix in LUKS funcs
+dnl and for guestfs_setfiles excludes parameter.
 PKG_CHECK_MODULES([LIBGUESTFS], [libguestfs >= 1.58.1])
 printf "libguestfs version is "; $PKG_CONFIG --modversion libguestfs
 
